src/matrix.rs

```diff
@@ -258,4 +341,141 @@ mod tests {
         let a = Mat4x4::identity();
         assert_eq!(a, a.transpose());
     }
+
+    #[test]
+    fn determinant_of_2x2_matrix() {
+        let a = Mat2x2::new([[1., 5.], [-3., 2.]]);
+        assert_eq!(17.0, a.determinant());
+    }
+
+    #[test]
+    fn submatrix_of_3x3() {
+        let a = Mat3x3::new([[1., 5., 0.], [-3., 2., 7.], [0., 6., -3.]]);
+        let expected = Mat2x2::new([[-3., 2.], [0., 6.]]);
+        assert_eq!(expected, a.submatrix(0, 2));
+    }
+
+    #[test]
+    fn submatrix_of_4x4() {
+        let a = Mat4x4::new([
+            [-6., 1., 1., 6.],
+            [-8., 5., 8., 6.],
+            [-1., 0., 8., 2.],
+            [-7., 1., -1., 1.],
+        ]);
+        let expected = Mat3x3::new([[-6., 1., 6.], [-8., 8., 6.], [-7., -1., 1.]]);
+        assert_eq!(expected, a.submatrix(2, 1));
+    }
+
+    #[test]
+    fn minor_of_3x3() {
+        let a = Mat3x3::new([[3., 5., 0.], [2., -1., -7.], [6., -1., 5.]]);
+        let b = a.submatrix(1, 0);
+        assert_eq!(25., b.determinant());
+        assert_eq!(25., a.minor(1, 0));
+    }
+
+    #[test]
+    fn cofactor_of_3x3() {
+        let a = Mat3x3::new([[3., 5., 0.], [2., -1., -7.], [6., -1., 5.]]);
+        assert_eq!(-12., a.minor(0, 0));
+        assert_eq!(-12., a.cofactor(0, 0));
+        assert_eq!(25., a.minor(1, 0));
+        assert_eq!(-25., a.cofactor(1, 0));
+    }
+
+    #[test]
+    fn determinant_of_3x3_matrix() {
+        let a = Mat3x3::new([[1., 2., 6.], [-5., 8., -4.], [2., 6., 4.]]);
+        assert_eq!(56., a.cofactor(0, 0));
+        assert_eq!(12., a.cofactor(0, 1));
+        assert_eq!(-46., a.cofactor(0, 2));
+        assert_eq!(-196., a.determinant());
+    }
+
+    #[test]
+    fn determinant_of_4x4_matrix() {
+        let a = Mat4x4::new([
+            [-2., -8., 3., 5.],
+            [-3., 1., 7., 3.],
+            [1., 2., -9., 6.],
+            [-6., 7., 7., -9.],
+        ]);
+        assert_eq!(690., a.cofactor(0, 0));
+        assert_eq!(447., a.cofactor(0, 1));
+        assert_eq!(210., a.cofactor(0, 2));
+        assert_eq!(51., a.cofactor(0, 3));
+        assert_eq!(-4071., a.determinant());
+    }
+
+    #[test]
+    fn invertible_matrix_is_invertible() {
+        let a = Mat4x4::new([
+            [6., 4., 4., 4.],
+            [5., 5., 7., 6.],
+            [4., -9., 3., -7.],
+            [9., 1., 7., -6.],
+        ]);
+        assert_eq!(-2120., a.determinant());
+        assert_eq!(true, a.invertible());
+    }
+
+    #[test]
+    fn noninvertible_matrix_is_not_invertible() {
+        let a = Mat4x4::new([
+            [-4., 2., -2., -3.],
+            [9., 6., 2., 6.],
+            [0., -5., 1., -5.],
+            [0., 0., 0., 0.],
+        ]);
+        assert_eq!(0., a.determinant());
+        assert_eq!(false, a.invertible());
+    }
+
+    #[test]
+    fn inverse_of_matrix() {
```

Comment:
The book has another test for inverse:
"Calculating the inverse of another matrix"

Reply:
Missed those, will add!